### PR TITLE
Use generalized profile fields for left sidebar

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [X] Feature

## Description

While we now support generalized profile fields (including a dynamically generated page for user's to update their profiles), we are not really using this information when displaying the profile, as was most recently pointed out by @AfroDevGirl [here](https://github.com/forem/forem/issues/6994#issuecomment-716571684)

As a quickfix, this PR makes the left sidebar of the profile dynamic, by rendering the fields with `display_area: :left_sidebar` there. This is **not** a complete overhaul of the profile page, only the sidebar, which is a step in the right direction. Maybe someone else can help out with the header, which I left untouched for now.

I also used Blazer to check the current profile fields on DEV, and 🎉 , all of them are assigned to `:left_sidebar`, which is definitely not what we want. So this PR also includes a data update script to assing all of the existing DEV fields to the correct areas and adds a new area, for fields that are only needed as settings. Two related questions:

1. Can we still do this via a data update script or will this negatively impact Forems which already modified their profiles? I can do a temporary Rake task instead that we run for DEV and that other communities can run if they choose too? CC @benhalpern 
2. Does it even make sense to have profile fields that are only used as settings? DId we over-generalize here? If they are configurable, we can't rely on them being there, which makes them not very useful as settings (e.g. the two color fields). Should these remain on `User` instead? That said, they currently also include things like `display_looking_for_work_on_profile` or `recruiters_can_contact_me_about_job_opportunities`, which are pretty DEV specific and might not apply to other communities at all.

## Related Tickets & Documents

Part of #6994

## QA Instructions, Screenshots, Recordings

1. Make sure you have some profile fields that are have a display area of `:left_sidebar`.
2. Run the code and make sure the profile page renders as you'd expect it to.

## Added tests?

- [X] Yes, but it's not working yet, so back to draft.

## Added to documentation?

- [ ] Docs.forem.com
- [ ] README
- [X] No documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

image

### CODE REVIEW

**Review by Ridhwana — APPROVED** [review]
This looks good as a first pass @citizen428. I've approved this pending Travis passing :)

**Comment by citizen428** [comment]
The failing spec has me stumped a bit. It keeps persistently failing, no matter which permuation of `has_content`/`has_text` and `within` I'm using. But if I look at the page (`save_and_open_page`) the content is right there:

Screen_Shot_2020-10-28_at_10_07_09

That said the failure screenshot generated by the spec looks like this

failures_r_spec_example_groups_user_index_when_user_is_unauthorized_when1_article_shows_all_proper_elements_299

and the error mentions the following towards the end:

> (However, it was found 1 time including non-visible text.)

**Review by mstruve — COMMENTED** [review]
> Can we still do this via a data update script or will this negatively impact Forems which already modified their profiles?

Thinking big picture here. It seems like this concern is going to come up a bit more before we finish this generalization project. What do you think about the idea of checking all the other Forems right now and figuring out if any have modified their profile fields? IF no one has modified them, we could take away that ability, go with defaults for everyone for the migration and then when we are done open up the existing fields for editing. Thoughts?

> Does it even make sense to have profile fields that are only used as settings? 

I agree "settings" dont feel like they fit perfectly into the profile fields model BUT I do think those more DEV specific ones you mention need to be configurable so leaving them on the User does not feel like the right solution. Option 1: move them to Profile fields so they are configurable and then maybe once everything is ironed out with the generalization we take a step back and assess whether we need another model. Option 2: we could leave them on User and also make the assessment after generalization is complete. I lean towards option 1 bc I think making them configurable is a step in the right direction even if the model definition doesn't fit perfectly. @vaidehijoshi you are good with this big picture stuff, any thoughts? 

> The failing spec has me stumped a bit.

Going to try to find some time to take a look at that

File: spec/system/user/view_user_index_spec.rb
Comment (mstruve) [inline spec/system/user/view_user_index_spec.rb:11]: This spec fails bc its in the "the user is unauthorized" context. If you move it down to the "when visiting own profile" context then it passes.

File: app/views/users/_sidebar.html.erb
Comment (vaidehijoshi) [reply app/views/users/_sidebar.html.erb:6]: +1 to extracting this out of the view and making the controller responsible for setting this (and passing it along to the view)

**Review by vaidehijoshi — COMMENTED** [review]
> > Does it even make sense to have profile fields that are only used as settings?
> 
> I agree "settings" dont feel like they fit perfectly into the profile fields model BUT I do think those more DEV specific ones you mention need to be configurable so leaving them on the User does not feel like the right solution. Option 1: move them to Profile fields so they are configurable and then maybe once everything is ironed out with the generalization we take a step back and assess whether we need another model. Option 2: we could leave them on User and also make the assessment after generalization is complete. I lean towards option 1 bc I think making them configurable is a step in the right direction even if the model definition doesn't fit perfectly. @vaidehijoshi you are good with this big picture stuff, any thoughts?

I think @mstruve's "option 1" suggestion is the right step here. I _do_ think we'll need another model to encapsulate some of these additional "profile fields". When I look at what we _currently_ have in `SETTINGS_FIELDS`, I see a few things that shouldn't _really_ be a part of a generic profile field, but instead, should be something that could be _added on_, if that makes sense. I almost envision the Forem creator being able to select "applicable profile fields" for their community.

For example, the `display_looking_for_work_on_profile`, `looking_for_work`, and `recruiters_can_contact_me_about_job_opportunities` settings [shown here](https://github.com/forem/forem/pull/11110/files#diff-8b23d24f18428c2c55b835c9b1ed1d37b9266d129b14eb921b14017bd6920b6cR36-R38) feel like they want to be a part of "hiring profile fields" option (almost like a widget or an add-on that a Forem creator could "apply" to ever user's profile field) that could be conditionally enabled by the Forem creator themselves.

All of this is to say, I think some of these fields want to be a subset of the default user profile fields, so extracting them is definitely the right step forward, but we should keep in mind that they likely won't live in there for long :)

File: spec/system/user/view_user_index_spec.rb
Comment (citizen428) [reply spec/system/user/view_user_index_spec.rb:11]: Yes, this is fixed locally, I just never pushed it because the PR went back to draft and I'm not sure if I'll continue working on it the current form or if a more comprehensive profile page revamp will be done right away.

File: app/views/users/_sidebar.html.erb
Comment (citizen428) [reply app/views/users/_sidebar.html.erb:6]: Normally, yes. 

For quick fixes that I don't expect to be around long, I often apply different rules: the fewer files I have to touch, the better. The logic (do we have a value and should we render something?) can not easily move into the controller and this draft PR was not the right place to come up with the proper abstraction. It was only meant to enable the Quidditch Forem to quickly finalize their profile setup and then be ripped out again ASAP. 

Now that we deployed the disabled feature flag for everyone, I have more time, so this will definitely move _somewhere_.

**Comment by citizen428** [comment]
This PR started out as a potential quick fix which I'm no longer sure we should pursue at all given that we deployed the feature flag for profile admin actions now. Thanks for the feeback though.

@mstruve 

> IF no one has modified them, we could take away that ability, go with defaults for everyone for the migration and then when we are done open up the existing fields for editing. 

Since this comment was written we deployed the feature flag PR, which disabled profile editing for all Forems. And as soon as we merge #10707 they'll also have the full set of DEV fields again. 

> I agree "settings" dont feel like they fit perfectly into the profile fields model BUT I do think those more DEV specific ones you mention need to be configurable so leaving them on the User does not feel like the right solution. 

This is why I initally moved them over. It was basically the "put them here until we find a better place for them" approach. Based on my living room this doesn't seem to work super well in the long term, so I thought it's time to start the discussion for the future.

@vaidehijoshi 

> For example, the display_looking_for_work_on_profile, looking_for_work, and recruiters_can_contact_me_about_job_opportunities settings shown here feel like they want to be a part of "hiring profile fields" option (almost like a widget or an add-on that a Forem creator could "apply" to ever user's profile field) that could be conditionally enabled by the Forem creator themselves.

I'm not sure if you watched my profile walkthrough video yet, but there I mention the mid to long-term idea of exporting/sharing/importing groups of related fields/settings. A lot of the groundwork for that is there (e.g. the CSV importer), but we'd still need to add export and import options to the admin section and a place to share these configurations. We've been taking a lot about plugins in the past, and I think this is what we really want here: an entity (the plugin) that is comprised of a number of settings plus the code utilizing them). But to the best of my knowledge, we haven't done any work on this yet, so it's more a long-term vision than an actual plan.

The real issue with all of this is that we still can't expect any profile field/setting to be present, so all code that tries to use them needs to guard against their absence and I really don't think litttering our codebase with conditionals is desirable.

**Comment by citizen428** [comment]
I'll close this for now. Parts will go into another PR, and then I'll decide if I rebase this and reopen it or do a completely new branch for a complete rework of the profile page.

**Pull request closed without merging**
